trace/trace.go

```diff
@@ -78,11 +109,10 @@ func (s traceStep) time() time.Time {
 	return s.stepTime
 }
 
-func (s traceStep) writeItem(b *bytes.Buffer, formatter string, startTime time.Time, stepThreshold *time.Duration) {
+func (s traceStep) writeItem(traceNum int32, startTime time.Time, stepThreshold *time.Duration) {
```

Comment:
this was intentionally constructed before to put all items related to a trace in a single log line... splitting them into per-item/per-step log lines is more expensive and makes it harder to find the information you need in a busy log

Reply:
I don't agree with this change making information less accessible. It's true that with multiple calls those logs can interweave with others, but it's strage to assume that users would not use tools like `grep` in this situation. Single traces can be easily extracted by greping for `Trace` and for `traceNum`.

From log ingestion standpoint, it's really strange that traces use some custom multiline format, that doesn't appear anywhere else. Using multilines is also super strage as most log processing is by lines (docker splits by lines, FluentD/FluentBit processes by lines). To make traces work you need to write custom regexpes/plugins to merge those multiple lines to just restore original information. This is the whole benefit of having structured logging where we can use easily parse-able metadata to correlate log lines. Having uniform log way to store information allows for community to develop common tools, instead of requiring admins to write their own regexp just to handle trace.

Performance should be addressed by implementation (e.g. switching to json would result in InfoS being 7 times more efficient then klog Infof with trace [more in KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging#logger-implementation-performance)). Of course we will work with scalability team to ensure that there is no performance regressions.

Reply:
> Performance should be addressed by implementation (e.g. switching to json would result in InfoS being 7 times more efficient then klog Infof with trace)

Is that still true even with the switch from a single call to Infof to per-step calls?

Reply:
My point is that we should not focus on performance from caller side and have optimizations done on log library side. We should not need to do string manipulation (like in trace case) on functions that already should be efficient.

Reply:
Yes, this was a hot path allocation.  This was i think 5% of all memory allocations on the server at one point, which is why this was fixed.

Reply:
Ok, It makes sense to avoid reintroducing same hot spot. We will try to propose a solution that will use only one log call and prepare some benchmarks. 

I guess Wojtek has most context about this.
/cc @wojtek-t

Reply:
> @hase1128 do you think you can rework your solution to use only one InfoS call?

OK, I will consider that. (all steps and nested trace need to be contained one []interface)

Reply:
I updated this PR(aggregated all trace information into one klog.InfoS).
Basically idea is same as original(I just changed bytes.buffer to []interface and modified related parts about that changing.)

Original log outputs
```
I0609 16:26:38.981874   28905 trace.go:201] Trace[1178003433]: "root" test-key:100 (09-Jun-2020 16:26:00.300) (total time: 681ms):
Trace[1178003433]: ---"step1" 100ms (16:26:00.400)
Trace[1178003433]: ---"step2" key-B:200,test-key:100,key-A:testtest 55ms (16:26:00.455)
Trace[1178003433]: ["nested" key-A:testtest 345ms (16:26:00.636)
Trace[1178003433]:  ---"nested-step1" test-key:100 110ms (16:26:00.746)]
Trace[1178003433]: ---"step3" key-A:testtest,key-B:200 -200ms (16:26:00.781)
Trace[1178003433]: [681.350356ms] [681.350356ms] END
```

After migration (this is one line and long.)
```
I0611 12:56:11.744206   11763 trace.go:186] "Trace" traceNum=625752080 traceName="root" test-key=100 startTime="2020-06-11 12:56:11.062297209 +0900 JST m=+67.507696689" totalTime="681.546055ms" traceType="step" traceNum=625752080 traceMsg="step1" durationTime="100.186817ms" startTime="2020-06-11 12:56:11.162484014 +0900 JST m=+67.607883506" traceType="step" traceNum=625752080 traceMsg="step2" key-B=200 test-key=100 key-A="testtest" durationTime="55.144899ms" startTime="2020-06-11 12:56:11.217628857 +0900 JST m=+67.663028405" Trace nested="start" traceMsg="nested" key-A="testtest" durationTime="345.761025ms" startTime="2020-06-11 12:56:11.398117565 +0900 JST m=+67.843517263" traceType="step" traceNum=625752080 traceMsg="nested-step1" test-key=100 durationTime="110.306092ms" startTime="2020-06-11 12:56:11.508423797 +0900 JST m=+67.953823355" Trace nested="end" traceType="step" traceNum=625752080 traceMsg="step3" key-A="testtest" key-B=200 durationTime="-200.309028ms" startTime="2020-06-11 12:56:11.543532272 +0900 JST m=+67.988931745" traceNum=625752080 endTime="681.546055ms" totalTime="681.546055ms"
```

in case of adding new line(this is not actual log after migration, I just describe for visibility during the review)
```
I0611 12:56:11.744206   11763 trace.go:186] "Trace" traceNum=625752080 traceName="root" test-key=100 startTime="2020-06-11 12:56:11.062297209 +0900 JST m=+67.507696689" totalTime="681.546055ms" 
traceType="step" traceNum=625752080 traceMsg="step1" durationTime="100.186817ms" startTime="2020-06-11 12:56:11.162484014 +0900 JST m=+67.607883506" 
traceType="step" traceNum=625752080 traceMsg="step2" key-B=200 test-key=100 key-A="testtest" durationTime="55.144899ms" startTime="2020-06-11 12:56:11.217628857 +0900 JST m=+67.663028405" 
Trace nested="start" traceMsg="nested" key-A="testtest" durationTime="345.761025ms" startTime="2020-06-11 12:56:11.398117565 +0900 JST m=+67.843517263" 
traceType="step" traceNum=625752080 traceMsg="nested-step1" test-key=100 durationTime="110.306092ms" startTime="2020-06-11 12:56:11.508423797 +0900 JST m=+67.953823355" 
Trace nested="end" 
traceType="step" traceNum=625752080 traceMsg="step3" key-A="testtest" key-B=200 durationTime="-200.309028ms" startTime="2020-06-11 12:56:11.543532272 +0900 JST m=+67.988931745" 
traceNum=625752080 endTime="681.546055ms" totalTime="681.546055ms"
```

remain tasks are here?
 - Each key&value' name is good or not(e.g. traceType="step", Trace nested="start")
 - Performance test (if necessary)
 - Fix trace_test.go (if necessary)